device/transport/mqtt/src/mqtt_twin_client.ts

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { EventEmitter } from 'events';
+import { TwinProperties } from 'azure-iot-device';
+import { MqttBase, translateError } from 'azure-iot-mqtt-base';
+import * as querystring from 'querystring';
+import * as url from 'url';
+import * as machina from 'machina';
+import * as dbg from 'debug';
+const debug = dbg('azure-iot-device-mqtt:MqttTwinClient');
+
+// $iothub/twin/PATCH/properties/reported/?$rid={request id}&$version={base version}
+
+/* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_009: [** The subscribed topic for `response` events shall be '$iothub/twin/res/#' **]** */
+const responseTopic = '$iothub/twin/res/#';
+
+/* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_019: [** The subscribed topic for post events shall be $iothub/twin/PATCH/properties/desired/# **]** */
+const desiredPropertiesUpdatesTopic = '$iothub/twin/PATCH/properties/desired/#';
+
+interface TopicSubscription {
+  mqttClient: MqttBase;
+  topic: string;
+}
+
+/**
+ * @private
+ * @class        module:azure-iot-device-mqtt.MqttTwinReceiver
+ * @classdesc    Acts as a receiver for device-twin traffic
+ *
+ * @param {Object} config   configuration object
+ * @fires MqttTwinReceiver#subscribed   an MQTT topic has been successfully subscribed to
+ * @fires MqttTwinReceiver#error    an error has occured while subscribing to an MQTT topic
+ * @fires MqttTwinReceiver#response   a response message has been received from the service
+ * @fires MqttTwinReceiver#post a post message has been received from the service
+ * @throws {ReferenceError} If client parameter is falsy.
+ *
+ */
+export class MqttTwinClient extends EventEmitter {
+  static desiredPropertiesUpdateEvent: string = 'twinDesiredPropertiesUpdate';
+
+  private _mqtt: MqttBase;
+  private _pendingTwinRequests: { [key: string]: any } = {};
+  private _topicFsm: machina.BehavioralFsm;
+  private _responseTopic: TopicSubscription;
+  private _desiredPropertiesUpdatesTopic: TopicSubscription;
+
+  constructor(client: MqttBase) {
+    super();
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_18_001: [The `MqttTwinClient` constructor shall accept a `client` object.]*/
+    this._mqtt = client;
+
+    const messageHandler = this._onMqttMessage.bind(this);
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_001: [The `MqttTwinClient` constructor shall immediately subscribe to the `message` event of the `client` object.]*/
+    this._mqtt.on('message', messageHandler);
+
+    this._topicFsm = new machina.BehavioralFsm({
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: function (client: TopicSubscription, err: Error, callback: (err?: Error) => void): void {
```

Comment:
Also confused.  Why is this called a "client" (a term that is so overly used in our code that it's meaningless), when I think "topic" might be a better name since this code is subscribing to a topic.

---
In reply to: [170588449](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170588449) [](ancestors = 170588449)

Reply:
I just got lazy and used whatever the machina.js doc used. I'll change it. topic sounds good.

---
In reply to: [170712943](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170712943) [](ancestors = 170712943,170588449)